client/lib/plugins/actions.js

```diff
@@ -64,11 +44,75 @@ function queueSitePluginAction( action, siteId, pluginId, callback ) {
 		} );
 	} else {
 		_actionsQueueBySite[ siteId ] = [];
-		action( siteId, pluginId, next.bind( this, callback ) );
+		action( next.bind( this, callback ) );
 	}
-}
+};
+
+const queueSitePluginActionAsPromise = ( action, siteId, pluginId, callback ) => {
+	var promise = new Promise( ( resolve, reject ) => {
+		queueSitePluginAction( action, siteId, pluginId, ( error, data ) => {
+			if ( callback ) {
+				callback( error, data );
+			}
+			if ( error ) {
+				reject( error );
+			}
+			resolve( data );
+		} );
+	} );
+
+	return promise;
+};
+
+const getSolvedPromise = ( dataToPass ) => {
+	return new Promise( resolve => resolve( dataToPass ) );
+};
+
+const getRejectedPromise = ( errorToPass ) => {
+	return new Promise( ( resolve, reject ) => reject( errorToPass ) );
+};
+
+/**
+ * Return plugin id depending if the site is a jetpack site
+ *
+ * @param {Object} site - site object
+ * @param {Object} plugin - plugin object
+ * @return {String} plugin if
+ */
+const getPluginId = ( site, plugin ) => {
```

Comment:
I think you're right but I'm not completely sure about that. For this reason I just did a refactoring which means that I haven't changed the whole behavior of this stuff.

I'm aware that plugins page for non-jetpack sites aren't using this code. Maybe we can update/modify the current approach but I prefer do it in another instance.